Source/qol/autopickup.cpp

```diff
@@ -30,31 +31,73 @@ bool HasRoomForGold()
 	return false;
 }
 
+int NumMiscItemsInInv(int iMiscId)
+{
+	int numItems = 0;
+	for (const auto &item : InventoryAndBeltPlayerItemsRange(Players[MyPlayerId])) {
+		if (item._iMiscId == iMiscId) {
+			numItems++;
+		}
+	}
+	return numItems;
+}
+
 } // namespace
 
-void AutoGoldPickup(int pnum)
+void AutoPickup(int pnum)
 {
-	if (!*sgOptions.Gameplay.autoGoldPickup)
-		return;
-
 	if (pnum != MyPlayerId)
 		return;
-	if (leveltype == DTYPE_TOWN)
-		return;
-	if (!HasRoomForGold())
+	if (leveltype == DTYPE_TOWN && !*sgOptions.Gameplay.autoPickupInTown)
 		return;
 
+	bool hasRoomForGold = HasRoomForGold();
+
 	for (auto pathDir : PathDirs) {
 		Point tile = Players[pnum].position.tile + pathDir;
 		if (dItem[tile.x][tile.y] != 0) {
 			int itemIndex = dItem[tile.x][tile.y] - 1;
 			auto &item = Items[itemIndex];
-			if (item._itype == ItemType::Gold) {
+			if (hasRoomForGold && item._itype == ItemType::Gold && *sgOptions.Gameplay.autoGoldPickup) {
 				NetSendCmdGItem(true, CMD_REQUESTAGITEM, pnum, pnum, itemIndex);
 				item._iRequest = true;
 			}
+			if (item._itype == ItemType::Misc && (AutoPlaceItemInInventory(Players[pnum], item, false) || AutoPlaceItemInBelt(Players[pnum], item, false))) {
+				bool doPickup = false;
+				switch (item._iMiscId) {
+				case IMISC_HEAL:
+					doPickup = sgOptions.Gameplay.numHealPotionPickup > NumMiscItemsInInv(item._iMiscId);
+					break;
+				case IMISC_FULLHEAL:
+					doPickup = sgOptions.Gameplay.numFullHealPotionPickup > NumMiscItemsInInv(item._iMiscId);
+					break;
+				case IMISC_MANA:
+					doPickup = sgOptions.Gameplay.numManaPotionPickup > NumMiscItemsInInv(item._iMiscId);
+					break;
+				case IMISC_FULLMANA:
+					doPickup = sgOptions.Gameplay.numFullManaPotionPickup > NumMiscItemsInInv(item._iMiscId);
+					break;
+				case IMISC_REJUV:
+					doPickup = sgOptions.Gameplay.numRejuPotionPickup > NumMiscItemsInInv(item._iMiscId);
+					break;
+				case IMISC_FULLREJUV:
+					doPickup = sgOptions.Gameplay.numFullRejuPotionPickup > NumMiscItemsInInv(item._iMiscId);
+					break;
+				case IMISC_ELIXSTR:
+				case IMISC_ELIXMAG:
+				case IMISC_ELIXDEX:
+				case IMISC_ELIXVIT:
+					doPickup = *sgOptions.Gameplay.autoElixirPickup;
+					break;
+				default:
+					break;
+				}

```

Comment:
This can be a bit more concise if extracted into a function (so you can do `return` instead of assignment + `break`)

Reply:
OK, I've done that.